modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpServerTransport.java

```diff
@@ -364,6 +366,18 @@ protected HttpMessage createMessage(String[] initialLine) throws Exception {
             }
             decoder.setCumulator(ByteToMessageDecoder.COMPOSITE_CUMULATOR);
             ch.pipeline().addLast("decoder", decoder); // parses the HTTP bytes request into HTTP message pieces
+
+            // from this point in pipeline every handler must call ctx or channel #read() when ready to process next HTTP part
+            ch.pipeline().addLast(new FlowControlHandler());
+            if (Assertions.ENABLED) {
+                // missing reads are hard to catch, but we can detect absence of reads within interval
+                long missingReadIntervalMs = 10_000;
+                ch.pipeline().addLast(new MissingReadDetector(transport.threadPool, missingReadIntervalMs));
+            }
+            // disable auto-read and issue first read, following reads must come from handlers
+            ch.config().setAutoRead(false);
+            ch.read();
```

Comment:
Likewise should we do this right at the end after setting up the whole pipeline?